internal/model/command.go

```diff
@@ -193,6 +218,151 @@ func (c ExecutableCommand[F]) GetFlagValues(cmd *cobra.Command) (*F, error) {
+	ctx := cmd.Context()
+	logger := log.From(ctx)
+
+	wf, wfPath, err := utils.GetWorkflow()
+	if err != nil {
+		return fmt.Errorf("failed to load workflow file: %w", err)
+	}
+
+	// If the workflow file doesn't exist, we're running locally, or the --pinned flag is set
+	// simply run the command normally with the existing version of the CLI
+	pinned, _ := cmd.Flags().GetBool("pinned")
+	if wf == nil || pinned || env.IsLocalDev() {
+		return nil
+	}
+
+	currentlyRunningVersion := events.GetSpeakeasyVersionFromContext(ctx)
+	artifactArch := ctx.Value(updates.ArtifactArchContextKey).(string)
+
+	// Try to migrate existing workflows
+	if wf.SpeakeasyVersion == "" {
+		if ghPinned := env.PinnedVersion(); ghPinned != "" {
+			wf.SpeakeasyVersion = workflow.Version(ghPinned)
+		} else {
+			wf.SpeakeasyVersion = "latest"
+		}
+
+		_ = updateWorkflowFile(wf, wfPath)
+	}
+
+	// Get the latest version, or use the pinned version
+	desiredVersion := wf.SpeakeasyVersion.String()
+	if desiredVersion == "latest" {
+		latest, err := updates.GetLatestVersion(artifactArch)
+		if err != nil {
+			return err
+		}
+		desiredVersion = latest.String()
+
+		logger.PrintfStyled(styles.DimmedItalic, "Running with latest Speakeasy version: %s\n", desiredVersion)
+	} else {
+		logger.PrintfStyled(styles.DimmedItalic, "Running with speakeasyVersion from workflow.yaml: %s\n", desiredVersion)
+	}
+
+	// If the desired version is the same as the currently running version of the CLI, just run the command
+	if desiredVersion == currentlyRunningVersion {
+		return nil
+	}
+
+	// Get lockfile version before running the command, in case it gets overwritten
+	lockfileVersion := getLockfileVersion()
+
+	runErr := runWithVersion(cmd, artifactArch, desiredVersion)
+	if runErr == nil {
+		// If the command failed to run with the latest version, try to run with the version from the lock file
+		if wf.SpeakeasyVersion == "latest" {
+			msg := fmt.Sprintf("Failed to run with Speakeasy version %s: %s\n", desiredVersion, "a") //TODO runErr.Error())
+			_ = log.SendToLogProxy(ctx, log.LogProxyLevelError, msg, nil)
+			logger.PrintfStyled(styles.DimmedItalic, msg)
+			if env.IsGithubAction() {
+				githubactions.AddStepSummary("# Speakeasy Version upgrade failure\n" + msg)
+			}
+
+			if lockfileVersion != "" {
+				logger.PrintfStyled(styles.DimmedItalic, "Rerunning with previous successful version: %s\n", lockfileVersion)
+				if err := runWithVersion(cmd, artifactArch, lockfileVersion); err != nil {
+					return err
+				} else {
+					os.Exit(0) // Exit to prevent the command from running twice
+				}
+			}
+		}
+
+		// If the command failed to run with the pinned version, fail normally
+		return runErr
+	}
+
+	// Exit to prevent the command from running twice
+	os.Exit(0)
+	return nil
+}
+
+func runWithVersion(cmd *cobra.Command, artifactArch, desiredVersion string) error {
+	vLocation, err := updates.InstallVersion(cmd.Context(), desiredVersion, artifactArch, 30)
```

Comment:
Couldn't desiredVersion already be installed here? Why would we be pulling it down again

Reply:
> For a user running locally wouldn't desiredVersion already likely be installed
No. Main use case is when pinned to a specific version, but even when on `latest` it's possible their CLI version won't be up to date. BUT, in the event that CLI version == desiredVersion, we bypass this logic altogether and just run normally